File: 2581-divide-players-into-teams-of-equal-skill/2581-divide-players-into-teams-of-equal-skill.java
class Solution {
    public long dividePlayers(int[] skill) {
        Arrays.sort(skill);
        int n=skill.length;
        long sumAns=0;
        int targetSum=0;
        targetSum=skill[0]+skill[n-1];
        int currSum=0;
        for(int i=0;i<n/2;i++){
           currSum=skill[i]+skill[n-i-1];
           if(currSum!=targetSum){
            return -1;
           }
           sumAns=sumAns+ (long)skill[i]*skill[n-i-1];
        }

        return sumAns;
    }
}

File: 2581-divide-players-into-teams-of-equal-skill/2581-divide-players-into-teams-of-equal-skill.cpp
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
    long long dividePlayers(vector<int>& skill) {
        int n = skill.size();
        int group = n / 2;
        long long sum = 0;
        
        unordered_map<int, int> ourmap;
        
        // Calculate total sum of skills and populate the frequency map
        for (int i = 0; i < n; i++) {
            ourmap[skill[i]]++;
            sum += skill[i];
        }
        
        // Each team's pair sum should be the total sum divided by the number of teams
        if (sum % group != 0) return -1;  // If the sum is not divisible, return -1
        int pairsum = sum / group;
        long long finalsum = 0;
        
        for (int i = 0; i < n; i++) {
            if (ourmap[skill[i]] == 0) continue;  // Skip if the current skill is already used
            
            int x = pairsum - skill[i];
            if (ourmap[x] > 0) {
                // Make sure we have enough players for the pair
                if (x == skill[i] && ourmap[x] < 2) {
                    return -1;  // If the skill and its complement are the same, we need at least 2
                }

                // Form the pair
                finalsum += (long long) skill[i] * x;
                ourmap[skill[i]]--;
                ourmap[x]--;
            } else {
                return -1;  // If we can't form a valid team, return -1
            }
        }
        
        return finalsum;
    }
};
